### DISCUSSION TITLE
[Dependency]why database dependency injection override does not work?

### DISCUSSION THREAD

**Discussion by wjs2063** [fastapi/fastapi#9548] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# yield db session
async def get_db():
    db = MongoClient("mongodb://111.111.1.1:54321/test")
    try:
        yield db.local
    finally:
        db.close()


# here is my API code

@apps.post('/api/user',response_model = List[Return_User_type])
async def create_users(user: User,db = Depends(get_db)):
    user = dict(user)
    user["password"] = pwd_context.hash(user["password"])
    db.user.insert_one(user)
    data = [db.user.find_one({"user_id":user["user_id"]})]
    serialize_objId(data)
    return JSONResponse(status_code = 202, content = data)

# then I want to reconfig test_db url 
# test_main.py is below 

async def override_get_db():
    db = MongoClient("mongodb://test_db_url:54254/test")
    try:
        yield db.local
    finally:
        db.close()
app.include_router(apps)
app.dependency_overrides[get_db] = override_get_db

# finally in terminal pytest 
# dependency overrides doesn't work 
# db_url is still 111.111.1.1  , not change test_db_url
```

### Description

I want to test API 

original db_url is 111.111.1.1  

for test I want to change db_url  test_db_url (example 222.222.2.2) 

So I override app.dependency_overrides[get_db] = overrides_get_db 

But db_url doest not change! 

I don't know what i missied it

### Operating System

Linux, macOS

### Operating System Details

Docker in macos 

container : centos:7

### FastAPI Version

0.73.0

### Python Version

Python 3.9.16

### Additional Context

.

**Comment by chrisK824** [comment]
From a quick look your code seems correct, have you tried updating to latest version of fastAPI?

Can you share your testing code as well?

**Reply by wjs2063** [reply]
````
from .main import app
from routes.routes import apps
from pymongo import MongoClient
from fastapi.testclient import TestClient
from .models.database import get_db


# change origin_db_url to  test_db url
async def override_get_db():
    db = MongoClient("mongodb://222.222.2.2:54254/test")
    try:
        yield db.local
    finally:
        db.close()
app.include_router(apps)
app.dependency_overrides[get_db] = override_get_db
client = TestClient(app)


def test_read_main():
    response = client.get('/')
    assert  response.status_code == 200


def test_user_create():
    response = client.post('/api/user',headers = {'accept': 'application/json',
                                                         'Content-Type': 'application/json'},
                           json = {
                               "name": "a",
                               "user_id": "b",
                               "email": "ccc@naver.com",
                               "password": "ddd",
                               "nickname": "fff",
                               "disabled": "False"
                           }
                           )
    assert response.status_code == 202
````

then I excueted command "pytest" in terminal 

But The test found that the db_url was still referencing 111.111.1.1 .

I shut down my origin_db_server , then I started test_db server . then  test was started 

but error code is " db_connection time out error to origin_db_server ".  it menas db_url does not change!

**Reply by wjs2063** [reply]
here is my main.py
````
from fastapi import FastAPI,Depends
from fastapi.staticfiles import StaticFiles
from routes.routes import apps
from fastapi.security import OAuth2PasswordBearer
from fastapi.middleware.cors import CORSMiddleware
from starlette.responses import FileResponse
from models.database import get_db

#app = FastAPI(dependencies=[Depends(get_db)])
app.mount("/static",StaticFiles(directory="static"),name = "static")
app.include_router(apps)

# lots of API code in apps 

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")


origins = [
    "*"
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.get("/")
async def main():
    return FileResponse("static/index.html")

````

**Reply by wjs2063** [reply]
i tried to update latest version of fastapi but still not change

**Reply by chrisK824** [reply]
Hey @wjs2063 , is that line indeed commented out or just missed it in the past here?

```
#app = FastAPI(dependencies=[Depends(get_db)])
```

also in your test file this line can be removed:
```
app.include_router(apps)
```

However, not really making the difference here I guess, haven't tested that out but to be honest seems good enough. Haven't tried the global dependency in FastAPI object myself, but the concept should remain the same. Just in case, try moving that dependency in the path function, if still doesn't work then try a minimal example to start ruling out hidden spots

**Reply by wjs2063** [reply, marked as answer]
Thank you so much for responding. Actually, I was working on a fastapi project in docker on top of m1 mac, but it seems to be an OS problem. When I changed the container OS image and reinstalled it, the code was applied. Thank you
